tex/chProgramming.tex

```diff
@@ -1192,6 +1208,108 @@ \subsection{Recursion on natural numbers}\label{ssec:recnat}
 notations introduced by the file.
 There is no better way to browse the library than reading these banners.}
 
+\section{Total functions}\label{sec:total}
+When discussing the subtraction function in Section~\ref{ssec:recnat},
+we were confronted with a striking difference between the usual
+mathematical approach to functions and the approach used in \Coq{} and
+the \mcbMC{} library.  In the latter context, {\em all functions are total
+  by construction}.  This is a difficulty when we want the language we
+use in the \mcbMC{} library to stick as close to possible to the
+language of usual mathematics, where some functions are {\em not
+  defined everywhere}.
+
+The functions from Section~\ref{ssec:recnat} where the difference is
+visible are function~\C{predn} and function~\C{subn}.  The former
+returns the predecessor of natural number, and intuitively, it should
+not be defined when this natural number is \C{O}, which has no
+predecessor.  The latter returns the subtraction of a number from
+another one, and intuitively it should not be defined when the first
+argument (from which subtraction is performed) is smaller than the
+amount being subtracted.
+
+Usual mathematical practice, learned by
+everybody from middle school, is that we can consider a bigger type of
+numbers, in which the natural numbers are injected implicitly, and
+where the predecessor of \C{0} and the results of exceptional
+subtractions do exist.  So the predecessor of 3 is a natural number,
+but the predecessor of \C{0} is not, it is a relative integer.  This
+versatility, allowed by mathematical practice, may seem to provide
+the necessary soupleness to be able to write mathematical formulas
+quickly and meaningfully.  It is not the case.  there are mathematical
+functions that are inherently not defined everywhere, in ways that
+cannot be recovered by just adding a new type.
+
+The example we will choose is the example of multiplicative inverse.
+For natural numbers, only \(1\) has a multiplicative inverse
+\(1^{-1}=1\), so that for every number \(n\), \(n \times 1 \times
+1^{-1} = n\).  In a suitably large numerical type, such as a type of
```

Comment:
`In a suitably large numerical type...`
But the cardinality of the set of natural numbers is the same as the cardinality of the rationals.
Maybe I misinterpreted this part.